OOModel/src/expressions/MetaCallExpression.cpp

```diff
@@ -62,4 +69,61 @@ MetaDefinition* MetaCallExpression::metaDefinition()
 	return ret;
 }
 
+void MetaCallExpression::bindMetaCalls(Model::Node* n, MetaBinding* binding)
+{
+	if (auto metaCall = DCast<MetaCallExpression>(n))
+		if (auto callee = DCast<ReferenceExpression>(metaCall->callee()))
+			for (auto i = 0; i < binding->mappings()->size(); i++)
+				if (callee->name() == binding->mappings()->at(i)->name())
+					callee->setName(binding->mappings()->at(i)->value()->name());
+
+	for (auto child : n->children()) bindMetaCalls(child, binding);
+}
+
+Declaration* MetaCallExpression::generate()
+{
+	// return cached generated tree if exists
+	if (generated()) return generated();
+
+	auto metaDef = metaDefinition();
+	if (!metaDef) return nullptr;
+
+	if (arguments()->size() != metaDef->arguments()->size())
+	{
+		qDebug() << "#metaDefArgs != #metaCallArgs";
+		return nullptr;
+	}
+
+	// prepare visitor arguments using meta call arguments
+	QMap<QString, Model::Node*> args;
+	for (auto i = 0; i < arguments()->size(); i++)
+		args.insert(metaDef->arguments()->at(i)->name(), arguments()->at(i));
+
+	// use meta bindings to add additional visitor arguments
+	for (auto i = 0; i < metaDef->metaBindings()->size(); i++)
+	{
+		auto metaBinding = metaDef->metaBindings()->at(i);
+
+		if (auto input = args[metaBinding->input()->name()])
+		{
+			auto cloned = input->clone();
+			bindMetaCalls(cloned, metaBinding);
+			args.insert(metaBinding->name(), cloned);
+		}
+	}
+
+	// stage 0: clone meta definition
+	// cache generated tree already at this point to enable reference resolution
```

Comment:
I find the way your write comments, extremely useful, keep doing it just like that :)